File: src/engines/WijesekaraEngine.ts
/**
 * WijesekaraEngine
 * Ported faithfully from org.fossify.keyboard.helpers.WijesekaraEngine
 * Maps standard hardware/software keys to traditional Sinhala Wijesekara layout characters.
 */

const directMap: Record<string, string> = {
  // Lowercase / unshifted
  'q': 'ු', 'w': 'අ', 'e': 'ැ', 'r': 'ර', 't': 'එ',
  'y': 'හ', 'u': 'ම', 'i': 'ස', 'o': 'ද', 'p': 'ච',
  'a': '්', 's': 'ි', 'd': 'ා', 'f': 'ෙ', 'g': 'ට',
  'h': 'ය', 'j': 'ව', 'k': 'න', 'l': 'ක', ';': 'ත',
  "'": '.', 'z': 'ූ', 'x': 'ං', 'c': 'ජ', 'v': 'ඩ',
  'b': 'ඉ', 'n': 'බ', 'm': 'ප', ',': 'ල', '.': 'ග',
  '/': '/',
  // Uppercase / shifted
  'Q': 'ූ', 'W': 'උ', 'E': 'ෑ', 'R': 'ඍ', 'T': 'ඔ',
  'Y': 'ශ', 'U': 'ඹ', 'I': 'ෂ', 'O': 'ධ', 'P': 'ඡ',
  'A': 'ෟ', 'S': 'ී', 'D': 'ෲ', 'F': 'ෆ', 'G': 'ඨ',
  'H': '්‍ය', 'J': 'ළු', 'K': 'ණ', 'L': 'ඛ', ':': 'ථ',
  'Z': 'ෳ', 'X': 'ඞ', 'C': 'ඣ', 'V': 'ඪ', 'B': 'ඊ',
  'N': 'භ', 'M': 'ඵ', '<': 'ළ', '>': 'ඝ', '?': '?'
};

export class WijesekaraEngine {
  static getSinhalaChar(ch: string): string {
    return directMap[ch] || ch;
  }

  static processKey(currentWord: string, newChar: string): { replacedCount: number; output: string } {
    const mapped = this.getSinhalaChar(newChar);
    return { replacedCount: 1, output: mapped };
  }
}

File: app/src/main/kotlin/org/fossify/keyboard/helpers/WijesekaraEngine.kt
package org.fossify.keyboard.helpers

object WijesekaraEngine {

    private val directMap = mapOf(
        'q' to "ු", 'w' to "අ", 'e' to "ැ", 'r' to "ර", 't' to "එ",
        'y' to "හ", 'u' to "ම", 'i' to "ස", 'o' to "ද", 'p' to "ච",
        'a' to "්", 's' to "ි", 'd' to "ා", 'f' to "ෙ", 'g' to "ට",
        'h' to "ය", 'j' to "ව", 'k' to "න", 'l' to "ක", ';' to "ත",
        '\'' to ".", 'z' to "ූ", 'x' to "ං", 'c' to "ජ", 'v' to "ඩ",
        'b' to "ඉ", 'n' to "බ", 'm' to "ප", ',' to "ල", '.' to "ග",
        '/' to "/",
        // Shifted layer
        'Q' to "ූ", 'W' to "උ", 'E' to "ෑ", 'R' to "ඍ", 'T' to "ඔ",
        'Y' to "ශ", 'U' to "ඹ", 'I' to "ෂ", 'O' to "ධ", 'P' to "ඡ",
        'A' to "ෟ", 'S' to "ී", 'D' to "ෲ", 'F' to "ෆ", 'G' to "ඨ",
        'H' to "්‍ය", 'J' to "ළු", 'K' to "ණ", 'L' to "ඛ", ':' to "ථ",
        'Z' to "ෳ", 'X' to "ඞ", 'C' to "ඣ", 'V' to "ඪ", 'B' to "ඊ",
        'N' to "භ", 'M' to "ඵ", '<' to "ළ", '>' to "ඝ", '?' to "?"
    )

    fun getSinhalaChar(ch: Char): String {
        return directMap[ch] ?: ch.toString()
    }

    fun processKey(currentWord: String, newChar: Char): Pair<Int, String>? {
        val mapped = getSinhalaChar(newChar)
        return Pair(1, mapped)
    }
}
